core/src/main/scala/org/apache/spark/sql/delta/RowId.scala

```diff
@@ -87,4 +90,120 @@ object RowId {
     }
     latestMetadata
   }
+
+  /**
+   * Iterator to assign fresh row IDs to AddFiles in an iterator of actions.
+   */
+  private class RowIdAssigningIterator(
+      actions: Iterator[Action],
+      highWatermark: Option[Long]) extends Iterator[Action] {
+
+    val oldHighWatermark: Long = highWatermark.getOrElse(-1L)
+    var newHighWatermark: Long = oldHighWatermark
+
+    override def hasNext: Boolean = actions.hasNext || (newHighWatermark != oldHighWatermark)
+
+    override def next(): Action = {
+      if (actions.hasNext) {
+        actions.next() match {
+          case a: AddFile if a.baseRowId.isEmpty =>
+            val baseRowId = newHighWatermark + 1L
+            if (a.numPhysicalRecords.isEmpty) {
+              throw new UnsupportedOperationException(
+                "Cannot assign row IDs without row count statistics.")
+            }
+            newHighWatermark += a.numPhysicalRecords.get
+            a.copy(baseRowId = Some(baseRowId))
+          case a => a
+        }
+      } else {
+        val watermarkToCommit = newHighWatermark
+        newHighWatermark = oldHighWatermark // reset high watermark to make hasNext return false
+        RowIdHighWaterMark(watermarkToCommit, preservedRowIds = false)
+      }
+    }
+  }
+
+  /**
+   * Assigns fresh row IDs to all AddFiles inside `actions` that do not have row IDs yet.
+   */
+  private[delta] def assignFreshRowIds(
+      spark: SparkSession,
+      protocol: Protocol,
+      snapshot: Snapshot,
+      actions: Iterator[Action]): Iterator[Action] = {
+    if (rowIdsAllowed(spark) && rowIdsSupported(protocol)) {
+      val highWatermark = extractHighWatermark(spark, snapshot).map(_.highWaterMark)
+      new RowIdAssigningIterator(actions, highWatermark)
+    } else {
+      actions
+    }
+  }
+
+  /**
+   * Reassigns newly assigned Row IDs (i.e. Row IDs above `readHighWaterMark`) of all AddFile
+   * actions in `actions`  to be above `winningHighWaterMarkOpt` (if it is defined), and
+   * updates the high watermark in `actions`.
+   */
+  private[delta] def reassignOverlappingRowIds(
+      readHighWaterMark: Long,
```

Comment:
```suggestion
      appliedHighWaterMark: Long,
```
It guess the most relevant thing here is that it was already applied / used for new baseRowIds in actions

Reply:
I kept `readRowIdHighWatermark` to be consistent with `CurrentTransactionInfo` where it does make a bit more sense to call it that way